### ISSUE TITLE
Usage of `Clone` is confusing

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#248]
This weirdness comes from:
1. Cloning a Module needs to have different tensor ids for backprop to work as expected, so cloning a tensor will change the tensor's id
2. Cloning is not implemented for Tensors with an OwnedTape (why not just clone the owned tape?)
3. `.duplicate()` must be used on tensors with an OwnedTape to get a copy of it without the tape

This is weird because:
1. it may not be clear why you can clone on some tensors (NoneTape) and not others (OwnedTape)
2. Unclear when to use duplicate vs clone
3. Cloning a tensor isn't actually a clone since it changes the id

See discussion starting at https://discord.com/channels/998570811650748417/998570812590260336/1031186211643084880

**Comment by coreylowman** [comment]
Proposal:
1. Clone on tensors should keep the id the same - If i clone a tensor, I expect it to refer to the same tensor in the graph
2. impl Clone for OwnedTape, so you can clone any tensor
3. Add something for "thin" cloning which will always produce a tensor with a NoneTape. perhaps this can be internal?
4. for Modules, either they need to override the clone impls to change the ids of the tensors, or they should use a different trait instead of `Clone`

**Comment by coreylowman** [comment]
@M1ngXU do you think its confusing if you can't clone a tensor with OwnedTape?

**Comment by coreylowman** [comment]
Actually, this could be a good use case for a `.detach()` method.

```rust
trait Tensor {
    ...
    /// Returns this tensor with a **new id** and no tape.
    fn detach(self) -> Self::NoTape;
}
```

So notably, the only way to get a new id on a tensor is to call detach (clone will always keep the same id)

**Comment by M1ngXU** [comment]
why is `detach` taking ownership?

**Comment by M1ngXU** [comment]
> @M1ngXU do you think its confusing if you can't clone a tensor with OwnedTape?

yes (i didn't even know this)

**Comment by coreylowman** [comment]
> why is detach taking ownership? if this is intented, something like without_tape is better?

Mainly to decouple cloning from detaching. Trying to make cloning explicit instead of implicit. Not sure if its worth it.

**Comment by M1ngXU** [comment]
> > why is detach taking ownership? if this is intented, something like without_tape is better?
> 
> ~Mainly to decouple cloning from detaching. Trying to make cloning explicit instead of implicit. Not sure if its worth it.~
> 
> Hmm actually I agree with you, what about this?
> 
> ```rust
> trait Tensor {
>     /// clones data & sets a new id
>     fn detached(&self) -> Self::NoTape;
> }
> ```

makes sense. but wouldn't `detached` be kinda the wrong name, as this would mean that you also get the tape?

**Comment by coreylowman** [comment]
Hmm yeah I think returning `Self` makes sense, especially since we already have `split_tape()` for returning NoTape and the Tape.

Any ideas for the name?

**Comment by M1ngXU** [comment]
> Hmm yeah I think returning `Self` makes sense, especially since we already have `split_tape()` for returning NoTape and the Tape.
> 
> Any ideas for the name?

what is the use of a new id? other gradients?

**Comment by coreylowman** [comment]
Mainly just having the same data but without the history for gradients. Now that you ask this though we can just capture this with:
```rust
// NOTE: a: T
let mut b: T::NoTape = TensorCreator::zeros();
b.mut_data().clone_from(a.data());
```

So I think you're right that we don't even need this as an extra function.

The only place this functionality is used is in [map_df_uses_fx](https://github.com/coreylowman/dfdx/blob/main/src/tensor_ops/utils.rs#L41)

**Comment by coreylowman** [comment]
Okay so in summary:
1. Removing duplicate
2. Cloning tensors will keep ids the same
3. Cloning modules will keep underlying tensor ids the same. You can just create a new module if you want ids to be different, and this may enable some nice things with computing gradients in multiple places
4. Cloning OwnedTape can be addressed separately. This currently isn't possible because GradientTape holds a Vec of `dyn FnOnce(...)`, and apparently you can't add Clone to a dyn object because it will make it object unsafe?

**Comment by M1ngXU** [comment]
> 4\. Cloning OwnedTape can be addressed separately. This currently isn't possible because GradientTape holds a Vec of `dyn FnOnce(...)`, and apparently you can't add Clone to a dyn object because it will make it object unsafe?

can this be converted to `Box<dyn>`? Can this be cloned then?

**Comment by M1ngXU** [comment]
ok interesting (never worked with dyn traits)
